python/dgl/graphbolt/graph_storage/csc_sampling_graph.py

```diff
@@ -304,6 +306,85 @@ def sample_neighbors(
+            Node pairs representing source-destination edges.
+            - If `node_pairs` is a tuple: It should be in the format ('u', 'v')
+            representing source and destination pairs.
+            - If `node_pairs` is a dictionary: The keys should be edge type and
+            the values should be corresponding node pairs.
+        Returns
+        -------
+        Tuple[node_pairs, CSCSamplingGraph]
+            The compacted node pairs and the created CSCSamplingGraph object.
+        """
+
+        nodes_dict = defaultdict(list)
+        unique_nodes_all_types = []
+        if self.metadata:
+            node_type_to_id = self.metadata.node_type_to_id
+        node_type_offset = self.node_type_offset
+        is_mapping = isinstance(node_pairs, Mapping)
+        # Treat it as a homogeneous graph.
+        if is_mapping:
+            assert (
+                node_type_to_id is not None and node_type_offset is not None
+            ), "Please note that sampling with different edge types is \
+                not supported in a homogeneous graph."
+        else:
+            node_type_to_id = {"_N": 0}
+            node_pairs = {("_N", "_E", "_N"): node_pairs}
+            node_type_offset = [0, self.num_nodes]
+
+        # Collect nodes for each node type.
+        for etype, node_pair in node_pairs.items():
+            u_type, _, v_type = etype
+            u, v = node_pair
+            nodes_dict[u_type].append(u)
+            nodes_dict[v_type].append(v)
+
+        compacted_nodes_dict = defaultdict(list)
+        for ntype, nodes in nodes_dict.items():
+            collected_nodes = torch.cat(nodes)
+            # Compact and find unique nodes.
+            unique_nodes, collected_nodes = torch.unique(
+                collected_nodes, return_inverse=True
+            )
+            ntype_id = node_type_to_id[ntype]
+            # Convert heterogeneous unqiue node id to homogeneous node id.
+            unique_nodes = unique_nodes + node_type_offset[ntype_id]
+            unique_nodes_all_types.append(unique_nodes)
+            compacted_nodes_dict[ntype] = collected_nodes
+
+        # Map back in same order as collect.
+        compacted_node_pairs = {}
+        for etype, node_pair in node_pairs.items():
+            u_type, _, v_type = etype
+            u, v = node_pair
+            u_size, v_size = u.numel(), v.numel()
+            u = compacted_nodes_dict[u_type][:u_size]
+            compacted_nodes_dict[u_type] = compacted_nodes_dict[u_type][u_size:]
+            v = compacted_nodes_dict[v_type][:v_size]
+            compacted_nodes_dict[v_type] = compacted_nodes_dict[v_type][v_size:]
+            compacted_node_pairs[etype] = (u, v)
+        if not is_mapping:
+            compacted_node_pairs = list(compacted_node_pairs.values())[0]
+        # Perform neighbor sampling.
+        unique_nodes = torch.cat(unique_nodes_all_types)
+        return (
+            compacted_node_pairs,
```

Comment:
After compacting the node pairs, what would be the dst nodes when we call `dgl.to_block`?

Reply:
The dst nodes should be the seed nodes in the sampled subgraph,